### ISSUE TITLE
Quarks not working in branch 3.8dev on Raspberry Pi 3

### ISSUE THREAD

**Issue by multivac61** [supercollider/supercollider#2178]
I just installed the newest branch as explained [here](http://supercollider.github.io/development/building-raspberrypi.html). When I however try to install Quarks on the system I am not able to use them without ERRORs.

I am for example able to install the _TestQuark_ quark by using the command `Quarks.install("testquark")`, then I recompile the class library by restarting `sclang`. Once that's finished I try to call the `TestQuark` class but it does not work:

```
sc3> TestQuark
ERROR: Class not defined.
  in file 'selected text'
  line 1 char 9:

  TestQuark 
           
-----------------------------------
```

**Comment by jamshark70** [comment]
What messages appear in the post window during installation?

The error you described is what would happen if quark installation failed. In that case, the post window will explain what went wrong. Without knowing this, there's really no way to address the problem.

**Comment by multivac61** [comment]
Of course. So for example when I try to install `UnitTesting` I get the following response:

```
sc3> Quarks.install("UnitTesting");
Installing UnitTesting
Cloning into '/home/pi/.local/share/SuperCollider/downloaded-quarks/UnitTesting'...
remote: Counting objects: 109, done.
remote: Total 109 (delta 0), reused 0 (delta 0), pack-reused 109
Receiving objects: 100% (109/109), 40.96 KiB | 0 bytes/s, done.
Resolving deltas: 100% (62/62), done.
Checking connectivity... done.
Adding path: /home/pi/.local/share/SuperCollider/downloaded-quarks/UnitTesting
UnitTesting installed
-> Quark: UnitTesting[2a8f5e18aaa7cd25968ec094bbfa75a7fb3094b6]
```

**Comment by jamshark70** [comment]
If you note here, the class is called `UnitTest` rather than UnitTesting. So we have two things now in this issue report:

1. A quark (UnitTesting) that was known to install successfully, but which was invoked with a misspelled identifier.

2. A quark (TestQuark) that we aren't totally sure installed successfully (but might well have -- the obvious reasons for failure would have broken the other install as well). This one was invoked with a correct name, but wasn't found.

Maybe try uninstalling testquark, reinstalling, and copy the post window output here.

**Comment by multivac61** [comment]
Yes, I misspelled the class name, but in this case typing in the correct class name `UnitTest` gives me the same output as before, namely:

```
sc3> UnitTest
ERROR: Class not defined.
  in file 'selected text'
  line 1 char 8:

  UnitTest 
          
-----------------------------------
-> nil
```

As for the testquark, the output is the following:

```
sc3> Quarks.uninstall("testquark")
-> Quarks
```

Then I recompile the class library and install again:

```
sc3> Quarks.install("testquark")
Installing testquark
Adding path: /home/pi/.local/share/SuperCollider/downloaded-quarks/testquark
testquark installed
-> Quark: testquark[60aa9d3045c1befe71f0ec3c5c49a68a45ad0cdc]
```

and recompile the class library once again and calling the `TestQuark` class:

```
sc3> TestQuark
ERROR: Class not defined.
  in file 'selected text'
  line 1 char 9:

  TestQuark 
           
-----------------------------------
-> nil
```

Still gives me the `ERROR: Class not defined.`
```

**Comment by jamshark70** [comment]
1. What's the post window output while compiling the class library?

2. Check the locations where it claimed to download the files: /home/pi/.local/share/SuperCollider/downloaded-quarks/testquark and /home/pi/.local/share/SuperCollider/downloaded-quarks/UnitTesting -- are they populated?

3. What is the result of `LanguageConfig.includePaths`?

**Comment by multivac61** [comment]
1. ```
compiling class library...
initPassOne started
        NumPrimitives = 591
initPassOne done
        compiling dir: '/usr/local/share/SuperCollider/SCClassLibrary'
        pass 1 done
ERROR: Class extension for nonexistent class 'Document'
     In file:'/deprecated/3.7/deprecated-3.7.sc'
        numentries = 661942 / 6739680 = 0.098
        3695 method selectors, 1824 classes
        method table size 4448960 bytes, big table size 26958720
        Number of Symbols 8861
        Byte Code Size 251743
        compiled 221 files in 0.50 seconds

Info: 3 methods are currently overwritten by extensions. To see which, execute:
MethodOverride.printAll
```

2. `/home/pi/.local/share/SuperCollider/downloaded-quarks/testquark` contains files `testQuark.quark` and `TestQuark.sc`. `/home/pi/.local/share/SuperCollider/downloaded-quarks/UnitTesting` also contains the right files. So yes the locations are populated.

3. ```sc3> LanguageConfig.includePaths
[  ]
```

**Comment by jamshark70** [comment]
Well, that's interesting.

When it says, "Adding path: /home/pi/.local/share/SuperCollider/downloaded-quarks/UnitTesting," it means adding to the include paths. That information is supposed to be saved into the language configuration file, but this is not happening for some reason.

Are there any files in ~/.config/SuperCollider? If so, what's in the sclang configuration file? I forget the exact name but it's sclang something dot yaml.

**Comment by multivac61** [comment]
So the problem is that the Quarks aren't being found and thus not compiled/linked?

There are not configuration files under `~/config/SuperCollider` present (the folder doesn't exist).

**Comment by jamshark70** [comment]
~/.config/SuperCollider -- don't forget the dot before config

**Comment by multivac61** [comment]
`~/.config/SuperCollider` is what I meant, I'm sorry for being so hasty.

**Comment by jamshark70** [comment]
Ah ok. So, sclang should be creating this folder, but it isn't. Yep, that's a bug.

So... Config gurus, why would sclang not create its config file when run from the command line? It's a little bit important because quarks depends on it...

**Comment by jamshark70** [comment]
Changing the title of this issue to reflect the real problem: "LanguageConfig.store fails silently if the config directory doesn't exist."

This is what is supposed to happen during quark installation:

1. If the quark isn't found locally, fetch it using git. (This is working in your environment.)

2. Add the quark paths to the LanguageConfig "includePaths" collection.

3. Save the updated LanguageConfig to disk.

Currently, it seems that we rely on the IDE to create the SuperCollider config folder. If you have only ever run sclang from the command line, then this folder is never created. `LanguageConfig.store` then fails silently -- it appears everything worked, but sclang_conf.yaml is never saved to disk, and then the next time you run sclang, the quark paths are not included in library compilation.

As a fix, I would suggest:

- If the user didn't pass a path to `store` and the default config directory doesn't exist, automatically create it.

- If the user *did* pass a path and that directory location doesn't exist, throw an error.

Confirmed by testing:

1. Rename my .config/SuperCollider and Extensions folders to something else.

2. sclang

3. `Quarks.install("wslib")`

4. `LanguageConfig.store`

5. In a file browser, check ~/.config -- the SuperCollider/ directory still doesn't exist.

6. `0.exit`

7. Manually create ~/.config/SuperCollider.

8. Repeat 2-4. Now sclang successfully creates sclang_conf.yaml.
